File: ios/Crypto/encryption/AesGcmEncryption.swift
import Foundation
import CryptoKit

/**
 * AES-256-GCM authenticated encryption.
 *
 * Output format: nonce (12 bytes) || ciphertext || authTag (16 bytes)
 * — identical to AES.GCM.SealedBox.combined and Android's AES/GCM/NoPadding output.
 * This layout guarantees cross-platform compatibility.
 */
@available(iOS 13.0, *)
enum AesGcmEncryption {
  static let ivLength = 12
  static let tagLength = 16
  private static let keyLength = 32

  /**
   * Encrypts plaintext and returns nonce || ciphertext || authTag.
   * key must be exactly 32 bytes (AES-256).
   */
  static func encrypt(plaintext: Data, key: Data) throws -> Data {
    try requireKeyLength(key)
    let symmetricKey = SymmetricKey(data: key)
    let nonce = try AES.GCM.Nonce(data: CryptoBytes.randomBytes(count: ivLength))
    let sealedBox = try AES.GCM.seal(plaintext, using: symmetricKey, nonce: nonce)
    guard let combined = sealedBox.combined else {
      throw NSError(domain: "AesGcmEncryption", code: 0, userInfo: [
        NSLocalizedDescriptionKey: "AES.GCM.seal did not produce a combined output",
      ])
    }
    return combined
  }

  /**
   * Decrypts payload structured as nonce (12 bytes) || ciphertext || authTag (16 bytes).
   */
  static func decrypt(payload: Data, key: Data) throws -> Data {
    try requireKeyLength(key)
    guard payload.count >= ivLength + tagLength else {
      throw NSError(domain: "AesGcmEncryption", code: 1, userInfo: [
        NSLocalizedDescriptionKey: "Ciphertext is too short to be valid AES-GCM output",
      ])
    }
    let symmetricKey = SymmetricKey(data: key)
    let sealedBox = try AES.GCM.SealedBox(combined: payload)
    return try AES.GCM.open(sealedBox, using: symmetricKey)
  }

  private static func requireKeyLength(_ key: Data) throws {
    guard key.count == keyLength else {
      throw NSError(domain: "AesGcmEncryption", code: 2, userInfo: [
        NSLocalizedDescriptionKey: "AES-256-GCM requires exactly 32-byte key; received \(key.count) bytes",
      ])
    }
  }
}

File: android/src/main/java/com/securitysuite/crypto/encryption/AesGcmEncryption.java
package com.securitysuite.crypto.encryption;

import com.securitysuite.crypto.utils.CryptoBytes;

import javax.crypto.Cipher;
import javax.crypto.spec.GCMParameterSpec;
import javax.crypto.spec.SecretKeySpec;

/**
 * AES-256-GCM authenticated encryption.
 *
 * Output format: IV (12 bytes) || ciphertext || authTag (16 bytes)
 * This layout is identical to the iOS CryptoKit AES.GCM.SealedBox.combined format,
 * ensuring cross-platform compatibility.
 */
public final class AesGcmEncryption {
  private static final String TRANSFORMATION = "AES/GCM/NoPadding";
  public static final int IV_LENGTH = 12;
  public static final int TAG_LENGTH_BITS = 128;
  public static final int TAG_LENGTH_BYTES = TAG_LENGTH_BITS / 8;
  private static final int KEY_LENGTH = 32;

  private AesGcmEncryption() {}

  /**
   * Encrypts plaintext and returns IV || ciphertext || authTag.
   * @param plaintext raw bytes to encrypt
   * @param keyBytes  AES-256 key — must be exactly 32 bytes
   */
  public static byte[] encrypt(byte[] plaintext, byte[] keyBytes) throws Exception {
    requireKeyLength(keyBytes);
    byte[] iv = CryptoBytes.randomBytes(IV_LENGTH);
    GCMParameterSpec params = new GCMParameterSpec(TAG_LENGTH_BITS, iv);
    Cipher cipher = Cipher.getInstance(TRANSFORMATION);
    cipher.init(Cipher.ENCRYPT_MODE, new SecretKeySpec(keyBytes, "AES"), params);
    byte[] ciphertextWithTag = cipher.doFinal(plaintext);
    return concat(iv, ciphertextWithTag);
  }

  /**
   * Decrypts payload structured as IV (12 bytes) || ciphertext || authTag (16 bytes).
   */
  public static byte[] decrypt(byte[] payload, byte[] keyBytes) throws Exception {
    requireKeyLength(keyBytes);
    if (payload.length < IV_LENGTH + TAG_LENGTH_BYTES) {
      throw new IllegalArgumentException("Ciphertext is too short to be valid AES-GCM output");
    }
    GCMParameterSpec params = new GCMParameterSpec(TAG_LENGTH_BITS, payload, 0, IV_LENGTH);
    Cipher cipher = Cipher.getInstance(TRANSFORMATION);
    cipher.init(Cipher.DECRYPT_MODE, new SecretKeySpec(keyBytes, "AES"), params);
    return cipher.doFinal(payload, IV_LENGTH, payload.length - IV_LENGTH);
  }

  private static void requireKeyLength(byte[] keyBytes) {
    if (keyBytes == null || keyBytes.length != KEY_LENGTH) {
      throw new IllegalArgumentException(
          "AES-256-GCM requires exactly 32-byte key; received "
              + (keyBytes == null ? "null" : keyBytes.length + " bytes")
      );
    }
  }

  private static byte[] concat(byte[] a, byte[] b) {
    byte[] result = new byte[a.length + b.length];
    System.arraycopy(a, 0, result, 0, a.length);
    System.arraycopy(b, 0, result, a.length, b.length);
    return result;
  }
}
